cpp/src/spatial/linestring_distance.cu

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuspatial/error.hpp>
+#include <cuspatial/types.hpp>
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/utilities/device_atomics.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+
+#include <limits>
+#include <memory>
+#include <type_traits>
+
+namespace cuspatial {
+namespace {
+
+/** @brief Get the index that is one-past the end point of linestring at @p linestring_idx
+ *
+ * @note The last endpoint of the linestring is not included in the offset array, thus
+ * @p num_points is returned.
+ */
+template <typename OffsetIterator>
+inline cudf::size_type __device__
+endpoint_index_of_linestring(cudf::size_type const& linestring_idx,
+                             OffsetIterator const& linestring_offsets_begin,
+                             cudf::size_type const& num_linestrings,
+                             cudf::size_type const& num_points)
+{
+  return (linestring_idx == (num_linestrings - 1)
+            ? (num_points)
+            : *(linestring_offsets_begin + linestring_idx + 1)) -
+         1;
+}
+
+/**
+ * @brief Computes shortest distance between @p C and segment @p A @p B
+ */
+template <typename T>
+double __device__ point_to_segment_distance(coord_2d<T> const& C,
+                                            coord_2d<T> const& A,
+                                            coord_2d<T> const& B)
+{
+  double L_squared = (A.x - B.x) * (A.x - B.x) + (A.y - B.y) * (A.y - B.y);
```

Comment:
Let's add vec_2d operators (+, -, dot, scalar *) and make this code more geometrically intuitive.

Also, change this function to `point_to_segment_distance_squared(...)`...

```
template <typename T>
double __device__ point_to_segment_distance_squared(coord_2d<T> const& c,
                                            coord_2d<T> const& a,
                                            coord_2d<T> const& b)
{
  auto ab = b - a;
  auto ac = c - a;
  auto l_squared = dot(ab, ab); // could also use operator*(vec, vec) for dot
  if (L_squared == 0) { return dot(ac, ac); }
  auto r = dot(ac, ab) / L_squared;
  if (r <= 0 or r >= 1) {
    auto bc = c - b;
    return std::min(dot(ac, ac), dot(bc, bc));
  }
  auto p = a + r * ab;
  auto pc = c - p;
  return dot(pc, pc);
}
```

Reply:
> auto l_squared = dot(ab, ab); // could also use operator*(vec, vec) for dot

I think dot is better - you need an extra line to add the components with `vec*`.